labs/optimizely-data-cli/oevents

```diff
@@ -0,0 +1,651 @@
+{
+  echo "$@" >&2
+}
+
+################################################################################
+# Prerequisites                                                                #
+################################################################################
+
+function Require()
+{
+  local requirement="$1"
+
+  if ! command -v "$requirement" &> /dev/null; then
+    ErrorEcho "Error: $CLI_NAME requires $requirement to run.  Please install it and try again."
+    exit 1
+  fi
+}
+
+
+################################################################################
+# Date and Time                                                                #
+################################################################################
+
+# TimestampMsToDateString
+# Convert a unix timestamp (in milliseconds) to a date string and return the result
+function TimestampMsToDateString()
+{
+  local timestamp_ms="$1"
+  local timestamp_s=$(( timestamp_ms / 1000 ))
+  if [[ $(uname) = 'Darwin' ]]; then
+    echo $(date -r "$timestamp_s")
+  else
+    echo $(date -d "@$timestamp_s")
+  fi
+}
+
+# IncrDay
+# Increment a YYYY-MM-DD date string by one day and return the result
+function IncrDay()
+{
+  local to_be_incremented="$1"
+  
+  # From https://stackoverflow.com/questions/25701265/how-to-generate-a-list-of-all-dates-in-a-range-using-the-tools-available-in-bash
+  if [[ $(uname) = 'Darwin' ]]; then
+    echo $(date -j -v+1d -f "%Y-%m-%d" $to_be_incremented "+%Y-%m-%d")
+  else
+    echo $(date -I -d "$to_be_incremented + 1 day")
+  fi
+}
+
+# AssertIsBeforeOrEqual
+# Given two dates, assert that the first is before or equal to the second
+function AssertIsBeforeOrEqual()
+{
+  d1="$1"
+  d2="$2"
+
+  if [[ $(uname) = 'Darwin' ]]; then
+    ts1=$(date -j -u -f "%Y-%m-%d" "$d1" "+%s")
+    ts2=$(date -j -u -f "%Y-%m-%d" "$d2" "+%s")
+  else
+    ts1=$(date -d "$d1" +"%s")
+    ts2=$(date -d "$d2" +"%s")
+  fi
+
+  if [ "$ts1" -gt "$ts2" ]; then
+    ErrorEcho "Error: invalid date range: $d1 - $d2"
+    exit 1
+  fi
+}
+
+# ComputeDateRange
+# Compute a date range between $start and $end, and store the result in $date_range
+function ComputeDateRange()
+{
+  date_range=()
+
+  if [ -z "${start:-}" ]; then
+    return
+  fi
+
+  if [ -z "${end:-}" ]; then
+    end=$start
+  fi
+
+  AssertIsBeforeOrEqual "$start" "$end"
+
+  loopend=$(IncrDay "$end")
+
+  d="$start"
+  while [ "$d" != "$loopend" ]; do 
+    date_range+=("$d")
+    d=$(IncrDay "$d")
+  done
+}
+
+################################################################################
+# Authentication                                                               #
+################################################################################
+
+# CheckAWSCredentialExpiration
+# Check if the $AWS_SESSION_EXPIRATION timestamp exists and is in the past. If so
+# set $authenticated to false
+function CheckAWSCredentialExpiration()
+{
+  if [ ! -z "${AWS_SESSION_EXPIRATION:-}" ]; then
```

Comment:
This expression can be simplified to:

`if [[ -n ${AWS_SESSION_EXPIRATION} ]]; then`

Reply:
Helpful. Note however that we need to use

```sh
if [[ -n ${AWS_SESSION_EXPIRATION:-} ]]; then
```

because we're using `set -u`.  If we don't include `:-` in the substitution the script will fail when `AWS_SESSION_EXPIRATION` is undefined.

Reply:
Fixed.